pkg/services/ngalert/store/proto_instance_database.go

```diff
@@ -0,0 +1,261 @@
+			RuleUID string `xorm:"rule_uid"`
+			Data    []byte `xorm:"data"`
+		}
+
+		rows := make([]compressedRow, 0)
+		if err := sess.SQL(s.String(), params...).Find(&rows); err != nil {
+			return fmt.Errorf("failed to query alert_rule_state: %w", err)
+		}
+
+		for _, row := range rows {
+			instances, err := decompressAlertInstances(row.Data)
+			if err != nil {
+				return fmt.Errorf("failed to decompress alert instances for rule %s: %w", row.RuleUID, err)
+			}
+
+			// Convert proto instances to model instances
+			for _, protoInstance := range instances {
+				modelInstance := alertInstanceProtoToModel(row.RuleUID, row.OrgID, protoInstance)
+				if modelInstance != nil {
+					// If FlagAlertingNoNormalState is enabled, we should not return instances with normal state and no reason.
+					if st.FeatureToggles.IsEnabled(ctx, featuremgmt.FlagAlertingNoNormalState) {
+						if modelInstance.CurrentState == models.InstanceStateNormal && modelInstance.CurrentReason == "" {
+							continue
+						}
+					}
+					alertInstances = append(alertInstances, modelInstance)
+				}
+			}
+		}
+
+		return nil
+	})
+
+	logger.Debug("ListAlertInstances completed", "instances", len(alertInstances))
+
+	return alertInstances, err
+}
+
+func (st ProtoInstanceDBStore) SaveAlertInstance(ctx context.Context, alertInstance models.AlertInstance) error {
+	st.Logger.Error("SaveAlertInstance called and not implemented")
+	return errors.New("save alert instance is not implemented for proto instance database store")
+}
+
+func (st ProtoInstanceDBStore) FetchOrgIds(ctx context.Context) ([]int64, error) {
+	orgIds := []int64{}
+
+	err := st.SQLStore.WithDbSession(ctx, func(sess *db.Session) error {
+		s := strings.Builder{}
+		params := make([]any, 0)
+
+		addToQuery := func(stmt string, p ...any) {
+			s.WriteString(stmt)
+			params = append(params, p...)
+		}
+
+		addToQuery("SELECT DISTINCT org_id FROM alert_rule_state")
+
+		if err := sess.SQL(s.String(), params...).Find(&orgIds); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	return orgIds, err
+}
+
+func (st ProtoInstanceDBStore) DeleteAlertInstances(ctx context.Context, keys ...models.AlertInstanceKey) error {
+	logger := st.Logger.FromContext(ctx)
+	logger.Error("DeleteAlertInstances called and not implemented")
+	return errors.New("delete alert instances is not implemented for proto instance database store")
+}
+
+func (st ProtoInstanceDBStore) SaveAlertInstancesForRule(ctx context.Context, key models.AlertRuleKeyWithGroup, instances []models.AlertInstance) error {
+	logger := st.Logger.FromContext(ctx)
+	logger.Debug("SaveAlertInstancesForRule called", "rule_uid", key.UID, "org_id", key.OrgID, "instances", len(instances))
+
+	alert_instances_proto := make([]*pb.AlertInstance, len(instances))
+
+	for i, instance := range instances {
+		alert_instances_proto[i] = alertInstanceModelToProto(instance)
+	}
+
+	compressedAlertInstances, err := compressAlertInstances(alert_instances_proto)
```

Comment:
wdyt about adding a metric for the store so we can see avg or median size of the rule state?

Reply:
Sounds good, it might be useful to have this. I'll add it in a follow-up PR